### PR TITLE
[SPARK-21957][SQL] Add current_user function

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Added the `current_user` function.

## How was this patch tested?

Added UTs.

### CODE REVIEW

**Comment by holdenk** [comment]
The Python side looks reasonable, maybe @marmbrus can share his view on adding this function in general?
In the mean time Jenkins OK to test.

**Comment by HyukjinKwon** [comment]
looks something is gone wrong with Jenkins command .. not sure why but looks we should just say "ok to test"?

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.spark.sql.types.{DataType, StringType}
+
+/**
+ * Return the user executing the current query.
+ *
+ * There is no code generation since this expression should get constant folded by the optimizer.
+ */
+@ExpressionDescription(
+  usage = """
+    _FUNC_() - Returns the current user executing the query.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala]: I do not think Spark has something like `SESSION_USER`. Thus, this impl is wrong.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  """,
+  since = "2.3.0")
+case class CurrentUser() extends LeafExpression with CodegenFallback {
+  override def foldable: Boolean = true
+  override def nullable: Boolean = false
+  override def dataType: DataType = StringType
+
+  /** Returns the result of evaluating this expression on a given input Row */
+  override def eval(input: InternalRow): Any = {
+    SparkContext.getActive.map(_.sparkUser).getOrElse("")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala]: This is from Spark Context. All the sessions share the same user names. I am afraid this value is misleading to our end users. cc @rxin

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala]: This is actually the user which executes the query. I agree that may be misleading that a user can connect to STS with his credentials and then the queries are run with a superuser, but this is how Spark works at the moment and the user which actually runs the query is the superuser. If we don't think to STS, in the other use cases the user which runs a query is the user who started the session.

**Comment by mgaido91** [comment]
the error the build job is failing for is:
```
core/src/main/scala/org/apache/spark/io/CompressionCodec.scala:24: imported LZ4BlockInputStream' is permanently hidden by definition of object LZ4BlockInputStream in package io
[warn] import net.jpountz.lz4.{LZ4BlockInputStream, LZ4BlockOutputStream}
[warn]                         ^
[error] 
core/src/main/scala/org/apache/spark/io/CompressionCodec.scala:120: type mismatch;
[error]  found   : Boolean
[error]  required: net.jpountz.lz4.LZ4FastDecompressor
[error]     new LZ4BlockInputStream(s, disableConcatenationOfByteStream)
[error]                                ^
[info] Boolean <: net.jpountz.lz4.LZ4FastDecompressor?
[info] false
[warn] one warning found
[error] one error found
[error] (core/compile:compileIncremental) Compilation failed
```
this is not to be related to this PR. I am checking the reason which introduced the problem. Any idea?
Thanks.

**Comment by HyukjinKwon** [comment]
The test failure with -9 in Jenkins is unknown up to my knowledge (although there have been few clues and assumptions so far). It should be not related with it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala (hunk 2 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala]: Sorry, I do not think we should support it until we provide session-specific user management.

**Comment by maropu** [comment]
It seems the failure is caused by the version mismatch `v1.3` v.s. `v1.4` in lz4-java (recently, I updated in https://github.com/apache/spark/pull/18883). I'm not sure this happens in jenkins though...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala]: Add this to https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala?

**Comment by mgaido91** [comment]
On local I solved the build error by running a `mvn clean`. As pointed out by @maropu , a PR removed the class and then the incremental compilation fails. I am not sure why this is happening on jenkins though...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala]: ok, thanks. Then is there any plan to support session-specific user management? If so, can I help somehow? Thanks.

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CurrentUser.scala]: So far, it is not in our plan.

**Comment by gatorsmile** [comment]
So far, session-specific user management is not part of our plan yet. This API is not useful before it. Could you please close this PR?

**Comment by mgaido91** [comment]
I think it could have been useful anyway (eg. a user inserting rows through STS who wants to store the name of the user which is connecting to STS and performing the operation), but I agree that these are very corner cases. I am closing this then. Thanks.

**Comment by dmoore247** [comment]
It would be great to see this revived. A lot of old school ETL uses a current_user() function like this for audit/lineage/provenance fields.

**Pull request closed without merging**